Userland/Libraries/LibLocate/LocateDB.cpp

```diff
@@ -0,0 +1,203 @@
+ * SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+ * CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY,
+ * OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#include <LibLocate/LocateDB.h>
+
+namespace Locate {
+
+LocateDB::LocateDB(String path, LocateDbMode mode)
+{
+    if (!Core::File::exists(locate_db_path)) {
+        auto file_handle = Core::File::open(locate_db_path, Core::IODevice::WriteOnly, 0640);
+        if (file_handle.is_error()) {
+            outln("Got error while trying to create locate db file: {}", file_handle.error());
+            VERIFY_NOT_REACHED();
+        }
+        file_handle.value()->close();
+        chown(locate_db_path, 0, 18);
+    }
+
+    m_file_handle = fopen(path.characters(), (mode == LocateDbMode::Write ? "wb" : "rb"));
+    fseek(m_file_handle, 0, SEEK_END);
+    m_file_size = ftell(m_file_handle);
+    fseek(m_file_handle, 0, SEEK_SET);
+}
+
+LocateDB::~LocateDB()
+{
+    fclose(m_file_handle);
+}
+
+void LocateDB::write_header()
+{
+    // Format: magic string + version.
+    String db_header = String::formatted("{}|{}", "locatedb", 1);
+    fwrite(db_header.characters(), sizeof(uint8_t), db_header.length() + 1, m_file_handle);
+}
+
+void LocateDB::write_directory(DirectoryInfo& dir_info)
+{
+    struct DirStartEndHeader dir_header {
+        .type = (uint8_t)ChunkType::DirectoryStart,
+        .db_id = dir_info.db_id,
+        .parent_db_id = dir_info.parent_db_id,
+        .path_size = dir_info.path.length() + 1
+    };
+    fwrite(bit_cast<char*>(&dir_header), sizeof(struct DirStartEndHeader), 1, m_file_handle);
+    fwrite(dir_info.path.characters(), sizeof(uint8_t), dir_info.path.length() + 1, m_file_handle);
+
+    for (auto& child : dir_info.children) {
+        ChildFileData cur_child {
+            .type = (uint8_t)(child->type == FileType::File ? ChunkType::File : ChunkType::Directory),
+            .db_id = child->db_id,
+            .parent_db_id = child->parent_db_id,
+            .user_owner = child->user_owner,
+            .group_owner = child->group_owner,
+            .st_mode = child->st_mode,
+            .name_size = child->name.length() + 1
+        };
+
+        fwrite(bit_cast<char*>(&cur_child), sizeof(struct ChildFileData), 1, m_file_handle);
+        fwrite(child->name.characters(), sizeof(uint8_t), child->name.length() + 1, m_file_handle);
+    }
+
+    dir_header.type = (uint8_t)ChunkType::DirectoryEnd;
+    fwrite(bit_cast<char*>(&dir_header), sizeof(struct DirStartEndHeader), 1, m_file_handle);
+    fwrite(dir_info.path.characters(), sizeof(uint8_t), dir_info.path.length() + 1, m_file_handle);
+}
+
+bool LocateDB::verify_header()
+{
+    char buffer[BUFSIZ];
+    size_t bytes_read = fread(&buffer, sizeof(char), strlen("locatedb|1") + 1, m_file_handle);
```

Comment:
As this is hardcoded a second time here, and then a third time below, can we make it a constant instead? Something like

```cpp
constexpr const StringView MAGIC_HEADER { "locatedb|1" };
```